Cleanup type inference errors

imciner2 opened the pull request:
I have been going through the code using `@code_warntype` and we have had numerous places in the package where the compiler was unable to perform type inference at compile time leading to inefficient dispatching at runtime. This PR contains the high level changes to the main parts of the code, but there are still some type stability problems lurking in the depths of the code (such as the OrthoMADS computations). The main thing I have done is convert the abstract types inside of the structures to be concrete types and make the main problem then be parametric on them. There were also some problems using non-Float64 numbers (such as Float32), so I cleaned up the typing inside the constraints where that seemed to be the largest problem.

I benchmarked this on the rosenbrock function and here are the times I saw:

Before:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial:
  memory estimate:  9.08 MiB
  allocs estimate:  167154
  --------------
  minimum time:     16.152 ms (0.00% GC)
  median time:      16.368 ms (0.00% GC)
  mean time:        17.082 ms (2.46% GC)
  maximum time:     27.947 ms (22.48% GC)
  --------------
  samples:          293
  evals/sample:     1
```

After:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial:
  memory estimate:  7.08 MiB
  allocs estimate:  88178
  --------------
  minimum time:     11.699 ms (0.00% GC)
  median time:      11.802 ms (0.00% GC)
  mean time:        12.262 ms (2.65% GC)
  maximum time:     18.065 ms (22.72% GC)
  --------------
  samples:          408
  evals/sample:     1
```

imciner2 committed:
Make the time variables explicit in type

The time() function always returns Float64, so the parameterization
here is not useful (also, it was causing the compiler to not infer
a proper type for the variable returned from the struct, leading to
the use of Any for its operations).

imciner2 committed:
Fix config type inference

The lack of specialization in the DSProblem struct led to
poor type inference in the compiler.

imciner2 committed:
Cleanup Constraints type

There was poor type inference for the floating point types,
and also incompatibility with anything other than Float64. These
changes fix the inference and also allow Float32 to work.

imciner2 committed:
Make the mesh stored in the config a concrete type

Storing an AbstractMesh means that the compiler can't understand
the type of mesh that is being used, and since the mesh type won't
be changing during the run, just use a concrete type.

imciner2 committed:
Make the search step a concrete type

Storing it as an abstract type in the problem means
the type inference of the search phase is not optimal
(it has to always do dynamic dispatch).

imciner2 committed:
Convert the poll phase to an abstract type

imciner2 committed:
Make the cache type concrete

EdwardStables commented:
This is really great. 

You mentioned that you've changed types in Status to Float64 as that is the return type of time. Are we able to guarantee that this behavior is true on non 64-bit systems?

imciner2 commented:
> You mentioned that you've changed types in Status to Float64 as that is the return type of time. Are we able to guarantee that this behavior is true on non 64-bit systems?

Yep, it looks like we get a Float64 even on 32-bit systems. I looked into the base code that provides time here: https://github.com/JuliaLang/julia/blob/master/base/libc.jl#L249, and it is always casting to Float64 on return. So we should be fine with hardcoding those time variables to be always Float64.

imciner2 committed:
Give proper types to empty arrays.

Just using [] creates arrays of type Any, which then
lead to type conversions being introduced into the code.

imciner2 committed:
Cleanup typing inside the OrthoMADS functions

imciner2 commented:
Ok, I have now gone through the OrthoMADS functions and cleaned up some of the typing in there, along with some other pieces that I missed in other parts (such as the empty array intialization - don't use `[]`, use `T[]` to ensure it is created with the correct type and not Any). I think this takes care of all the type instability, all that is left would be excessive memory allocations and any cache inefficiencies.

Here are the timings now:

## LTMADS

Before:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial: 
  memory estimate:  9.08 MiB
  allocs estimate:  167092
  --------------
  minimum time:     16.146 ms (0.00% GC)
  median time:      16.279 ms (0.00% GC)
  mean time:        16.924 ms (2.67% GC)
  maximum time:     23.472 ms (15.51% GC)
  --------------
  samples:          296
  evals/sample:     1
```

After:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial: 
  memory estimate:  6.88 MiB
  allocs estimate:  86156
  --------------
  minimum time:     11.340 ms (0.00% GC)
  median time:      11.477 ms (0.00% GC)
  mean time:        11.970 ms (2.64% GC)
  maximum time:     21.027 ms (0.00% GC)
  --------------
  samples:          418
  evals/sample:     1
```

## OrthoMADS

Before:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3; poll=OrthoMADS()); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial: 
  memory estimate:  11.55 MiB
  allocs estimate:  243791
  --------------
  minimum time:     18.431 ms (0.00% GC)
  median time:      18.568 ms (0.00% GC)
  mean time:        19.391 ms (2.49% GC)
  maximum time:     28.367 ms (8.42% GC)
  --------------
  samples:          258
  evals/sample:     1
```

After:
```julia
julia> @benchmark Optimize!(p) setup=(p = DSProblem(3; poll=OrthoMADS()); SetObjective( p, DS.rosenbrock ))
BenchmarkTools.Trial: 
  memory estimate:  8.34 MiB
  allocs estimate:  98236
  --------------
  minimum time:     12.076 ms (0.00% GC)
  median time:      12.223 ms (0.00% GC)
  mean time:        12.688 ms (2.39% GC)
  maximum time:     22.559 ms (8.14% GC)
  --------------
  samples:          394
  evals/sample:     1
```

imciner2 committed:
Merge branch 'dev' into im/typechanges

EdwardStables commented:
It is likely that a fair few optimisations can be made in OrthoMADS, it was an almost direct implementation of the paper. But that could belong in another pull request. I think this is fine to merge?
